src/pytest_container/container.py

```diff
@@ -41,6 +41,75 @@ def __str__(self) -> str:
         return "oci" if self == ImageFormat.OCIv1 else "docker"
 
 
+@enum.unique
+class NetworkProtocol(enum.Enum):
+    """Network protocols support for port forwarding."""
+
+    #: Transmission Control Protocol
+    TCP = "tcp"
+    #: User Datagram Protocol
+    UDP = "udp"
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class PortForwarding:
+    container_port: int
+    host_port: int = -1
+    protocol: NetworkProtocol = NetworkProtocol.TCP
+
+    @property
+    def forward_cli_args(self) -> List[str]:
+        return [
+            "-p",
+            f"{self.host_port}:{self.container_port}/{self.protocol}",
+        ]
+
+    def __str__(self) -> str:
+        return str(self.forward_cli_args)
+
+
+def is_socket_listening_on_localhost(
+    port: int, protocol: NetworkProtocol
+) -> bool:
+    """Checks if a socket is listening on localhost for the given port and
+    protocol."""
+
+    open_ports = [
+        int(sock[6:].split(":")[-1])
+        for sock in LOCALHOST.socket.get_listening_sockets()
+        if sock[:3] == str(protocol)
+    ]
+    return port in open_ports
+
+
+def set_port_forwards(
+    port_forwards: List[PortForwarding],
+) -> List[PortForwarding]:
+    START_PORT = 1025
+    MAX_PORT = 2**16
+    finished_forwards: List[PortForwarding] = []
+
+    for port in port_forwards:
+        for i in range(START_PORT, MAX_PORT):
```

Comment:
while I see the convenience of such a check, iterating for each forward through 64512 ports is also expensive and I wonder, if the performance malus is really worth the convenience of preventing accidentally adding a port forward for an already occupied port. WDYT?

Reply:
if you add a port forward for an already occupied port, then your test will just fail and you have no way of influencing that besides restarting the test and hoping that the algorithm will not pick an occupied one next time. That's pretty bad UX imho.